### DISCUSSION TITLE
get test title in After() or Scenario()

### DISCUSSION THREAD

**Discussion by adudyak** [codeceptjs/CodeceptJS#3662] [Q&A] [answered]
Is there a way to access Test title in After() or at least in Scenario()?
`After(() => {
  this.test.title;
});`

I'm sending requests to API in scenarios, so would like to have a copy of sent request to be saved as file name %Scenario%.json

**Comment by kobenguyent** [comment]
I think those would help https://codecept.io/internal-api/#event-listeners

**Reply by adudyak** [reply]
Sample would be really appreciated. Got thorough all docs already.

**Reply by myrepojuly** [reply]
Hello, the code is the same as shared in the above link.

Here is an example:

You can create a file (say events.js). Inside the file, you can do as shown below

```
const { event } = require('codeceptjs');

module.exports = function() {

  event.dispatcher.on(event.test.before, function (test) {
    //play with the test obj

    console.log(test.title);

  });

  event.dispatcher.on(event.test.after, function (test) {

    //play with the test obj

    console.log(test.title);

  });
}
```
Then in the `codecept.conf.js` file, inside the helpers, add the below
 ```
 CustomSteps: {
      require: './events.js' //relative path to the events.js file
    }
```

**Reply by adudyak** [reply]
I see, "test" object is passed by framework itself, I just need to add my instructions in custom steps. Thank you, will try.

**Reply by myrepojuly** [reply]
In the `events.js` file, you can also store the `test.title` and other values like `test.tags`, 'test.parent'etc by capturing these in the `event.test.before` event and use it anywhere in your test framework by exporting it.

**Reply by adudyak** [reply]
@myrepojuly1987, do you have example in regular Scenario()?

**Reply by myrepojuly** [reply, marked as answer]
@adudyak -  You mean how to call the test title inside the test scenario? You can try something like below

```
const { event } = require('codeceptjs');

const testDetails = {
  title: null
};

event.dispatcher.on(event.test.before, function (test) {
  testDetails.title =test.title ;
});

module.exports = { testDetails };
```

Then in your test file,

```
const  {testDetails} = require('./events');

Feature('Test something');

Scenario('test scenario', async ({ I }) => {
  console.log(testDetails.title);
});
```
I don't think we the below is required in the `codecept.conf.js`. You can remove it, if the above works. Please test it.

```
CustomSteps: {
     require: './events.js' //relative path to the events.js file
   }
```

**Reply by myrepojuly** [reply]
Also, to avoid potential data races, you can create a new object 

```
const testDetails = [];

event.dispatcher.on(event.test.before, function (test) {
  testDetails.push({ title: test.title });
});
```

Then access the above in your tests

**Reply by adudyak** [reply]
got the approach, thanks

**Reply by adudyak** [reply]
@myrepojuly1987 I tried, it works. simply put that code on top level of test:
```
import { event } from 'codeceptjs';
event.dispatcher.on(event.test.before, function (test) {
  console.log(test.title);
});
```
Will see if I need any wrappers to hide code.
Thank you!
